engine/src/main/java/org/hibernate/validator/internal/metadata/aggregated/CascadingMetaDataBuilder.java

```diff
@@ -211,12 +195,57 @@ public CascadingMetaDataBuilder merge(CascadingMetaDataBuilder otherCascadingTyp
 	public CascadingMetaData build(ValueExtractorManager valueExtractorManager, Object context) {
 		validateGroupConversions( context );
 
-		if ( containerElementTypesCascadingMetaData.isEmpty() ) {
+		// we are 100% sure that we have a container and its elements are marked for cascading
+		if ( !containerElementTypesCascadingMetaData.isEmpty() && hasContainerElementsMarkedForCascading ) {
+			return ContainerCascadingMetaData.of( valueExtractorManager, this, context );
+		}
+
+		// if cascading == false we just return a non container metadata as we shouldn't be looking for
+		// any other cases like @Valid List list; or @Valid PotentialRuntimeContainer maybeContainerMaybeNot;
+		if ( !cascading ) {
 			return NonContainerCascadingMetaData.of( this, context );
 		}
-		else {
-			return ContainerCascadingMetaData.of( valueExtractorManager, this, context );
+
+		// now we are at the case when we might have something like @Valid SomeType something;
+		// First we should check if there are any available VE that can be applied to the current declared type (SomeType)
+		// If such are found we promote the cascading to the type variables that we have VEs for.
+		// In case of the Map call to valueExtractorManager.getPossibleValueExtractorCandidatesForCascadedValidation()
+		// will not return MapKeyExtractor, hence by default it will only return MapValueExtractor which matches BV Spec.
+		//
+		// If we found more than one type variable and an extractor for it we convert the metadata to something that we would
+		// have if the declaration was SomeContainer<@Valid VarWithVE1, @Valid VarWithVE2, VarWithoutVE ..., >
```

Comment:
OK, this is wrong. We need to throw an exception in this case as I mentioned in an earlier comment.